torch/csrc/jit/codegen/cuda/transform_iter.cpp

```diff
@@ -440,6 +473,282 @@ int BestEffortReplay::findFirstMismatchedID(
+        //   to both consumer and producer to the output of the expression
+        std::vector<IterDomain*> forwarded_ids;
+        std::vector<IterDomain*> compliment_ids;
+
+        for (auto input_id :
+             ir_utils::filterByType<IterDomain>(expr->inputs())) {
+          if (consumer_bcast_ids_not_in_producer.find(input_id) ==
+              consumer_bcast_ids_not_in_producer.end()) {
+            forwarded_ids.emplace_back(input_id);
+            forwarding_map.emplace(std::make_pair(input_id, merge_expr->out()));
+          } else {
+            compliment_ids.push_back(input_id);
+          }
+        }
+
+        // Set up compliment map
+        for (auto forwarded_id : forwarded_ids) {
+          compliment_map.emplace(std::make_pair(forwarded_id, compliment_ids));
+        }
+      }
+    }
+  }
+};
+
+} // namespace
+
+BestEffortReplay BestEffortReplay::replayCasP(
+    const TensorView* consumer,
+    const TensorView* producer,
+    int producer_compute_at_axis,
+    const RootDomainMap& root_map) {
+  if (producer_compute_at_axis < 0)
+    producer_compute_at_axis += (int)producer->nDims() + 1;
+
+  TORCH_INTERNAL_ASSERT(
+      producer_compute_at_axis >= 0 &&
+          (unsigned int)producer_compute_at_axis <= producer->nDims(),
+      "Invalid axis provided to BestEffortReplay::replayCasP.");
+
+  // Copied from TransformReplay::replayCasP
+  // TODO: Remove duplicate code
+
+  // producer ids we need to match in consumer
+  std::vector<IterDomain*> producer_CA_ids(
+      producer->domain()->domain().begin(),
+      producer->domain()->domain().begin() + producer_compute_at_axis);
+  producer_CA_ids = TensorDomain::noReductions(producer_CA_ids);
+
+  // If producer has an rfactor root, that's what will match the consumer
+  std::vector<IterDomain*> producer_root = producer->getMaybeRFactorDomain();
+
+  // Figure out all inputs required to generate the compute_at dimensions. We
+  // need all deps because inputs on producer may be in getRootDomain, but we
+  // may need in rFactorDomain
+  auto all_CA_id_deps = DependencyCheck::getAllValsBetween(
+      {producer_root.begin(), producer_root.end()},
+      {producer_CA_ids.begin(), producer_CA_ids.end()});
+
+  // Figure out minimal set of root IDs needed to produce producer_CA_ids:
+  std::unordered_set<IterDomain*> producer_CA_root_ids;
+  for (IterDomain* id : producer_root) {
+    if (std::find(all_CA_id_deps.begin(), all_CA_id_deps.end(), id) !=
+        all_CA_id_deps.end()) {
+      producer_CA_root_ids.emplace(id);
+    }
+  }
+
+  const auto p2c_root_map = root_map.mapProducerToConsumer(
+      producer->domain(), consumer->domain(), producer_CA_root_ids);
+
+  ConsumerForwardingInfo consumer_forwarding_info(producer, consumer);
+
+  auto consumer_replay = BestEffortReplay(
+      consumer->domain()->domain(),
+      producer_CA_ids,
+      p2c_root_map,
+      consumer_forwarding_info.forwarding_map);
```

Comment:
Why passing the consumer forwarding info here? This is CasP, so forwarding should never occur, right?

I added printf to see if forwarding actually occurs in this case. None of the C++ tests seems to do forwarding in this case, which seems to make sense to me.

Reply:
Forwarding during CasP is almost exclusively the entire PR, so I really hope so ;-) I'll add a test to demonstrate.

Reply:
Test case added.

Reply:
As far as I see, `consumer_forwarding_info.forwarding_map` is never used in this call of `BestEffortReplay`. See my change below. This is how I'm testing the behavior of these functions.

https://github.com/csarofeen/pytorch/pull/845/files#diff-7f84aefbee221c2e0b8ded6176bffc94c1c5c46304e8d30e510a9be56ae28855R288-R290

If it is indeed used, the assertion should fail, but none of the C++ tests, including the new ones, seem just fine.

The end results make sense to me. It just feel odd to pass `consumer_forwarding_info` as it should never be used.

Reply:
`consumer_forwarding_info` is indeed only supposed to be used for the consumer. `BestEffortReplay` called from `replayCasP` uses the forwarding info on the replay domain as that's where the consumer is. e.g.
```
replay_inps[t_i] = getForwardedId(it->second);
```

`BestEffortReplay` called from `replayPasC` uses forwarding info on the target domain. e.g.
```
target2replay_id_map_[getForwardedId(target_inp)] =  target2replay_it->second;
```

It's a bit of a clunky interface, but I wasn't sure where to draw the line on reusable interface vs duplicate code in `BestEffortReplay`.

Reply:
Compliment map is only supposed to be used with `replayCasP` as that's where we need to expand leaf nodes. If consumer is the target, we can simply ignore those since we're not creating a new domain based on them.

Reply:
Ah, I missed `replay_inps[t_i] = getForwardedId(it->second);`. Sorry.